File: projects/28-telecom-tower-monitoring/infra/burst.js
// Burst-load the queue to show backend scaling; synthetic signal names keep the live board partitions untouched.
import {
  SQSClient,
  GetQueueUrlCommand,
  SendMessageCommand,
  GetQueueAttributesCommand,
} from "@aws-sdk/client-sqs";

const ENDPOINT = process.env.AWS_ENDPOINT_URL || "http://localhost:4580";
const REGION = process.env.AWS_REGION || "eu-west-1";
const QUEUE = process.env.SQS_QUEUE_NAME || "ctm-tower-agg";

const args = process.argv.slice(2);
const opt = (name, def) => {
  const i = args.indexOf(name);
  return i >= 0 ? Number(args[i + 1]) : def;
};
const MESSAGES = opt("--messages", 2000);
const WORKERS = opt("--workers", 32);

const sqs = new SQSClient({
  endpoint: ENDPOINT,
  region: REGION,
  credentials: { accessKeyId: "test", secretAccessKey: "test" },
});
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

function body(i) {
  return JSON.stringify({
    sensor_type: `loadtest_${i % 5}`,
    site_id: `probe-${i % 20}`,
    window_start: "s",
    window_end: `e${i}`,
    count: 1, min: 0, max: 0, mean: 0, last: 0, spread: 0, alerts: [],
  });
}

async function depth(url) {
  const a = (await sqs.send(new GetQueueAttributesCommand({
    QueueUrl: url,
    AttributeNames: ["ApproximateNumberOfMessages", "ApproximateNumberOfMessagesNotVisible"],
  }))).Attributes || {};
  return Number(a.ApproximateNumberOfMessages || 0) + Number(a.ApproximateNumberOfMessagesNotVisible || 0);
}

async function main() {
  const url = (await sqs.send(new GetQueueUrlCommand({ QueueName: QUEUE }))).QueueUrl;
  const start = Date.now();
  let next = 0;
  async function worker() {
    while (next < MESSAGES) {
      const i = next++;
      await sqs.send(new SendMessageCommand({ QueueUrl: url, MessageBody: body(i) }));
    }
  }
  await Promise.all(Array.from({ length: WORKERS }, worker));
  console.log(`sent ${MESSAGES} messages in ${((Date.now() - start) / 1000).toFixed(2)}s`);

  const deadline = Date.now() + 120000;
  while (Date.now() < deadline) {
    const remaining = await depth(url);
    console.log("queue depth:", remaining);
    if (remaining === 0) {
      console.log("PASS: queue fully drained by the consumer");
      return;
    }
    await sleep(3000);
  }
  console.log("WARNING: queue not fully drained within the window (LocalStack throughput ceiling)");
}

main();

File: projects/27-river-flood-monitoring/infra/burst.py
"""Burst-load the queue to demonstrate backend scaling; synthetic signal names keep the live dashboard partitions untouched."""
import argparse
import json
import os
import time
from concurrent.futures import ThreadPoolExecutor

import boto3

ENDPOINT = os.getenv("AWS_ENDPOINT_URL", "http://localhost:4579")
REGION = os.getenv("AWS_REGION", "eu-west-1")
QUEUE = os.getenv("SQS_QUEUE_NAME", "rfw-catchment-agg")

_sqs = boto3.client("sqs", endpoint_url=ENDPOINT, region_name=REGION,
                    aws_access_key_id="test", aws_secret_access_key="test")


def _body(index):
    return json.dumps({
        "sensor_type": f"loadtest_{index % 5}", "site_id": f"probe-{index % 20}",
        "window_start": "s", "window_end": f"e{index}", "count": 1,
        "min": 0, "max": 0, "avg": 0, "latest": 0, "rise_mph": 0, "alerts": [],
    })


def _depth(url):
    attrs = _sqs.get_queue_attributes(QueueUrl=url, AttributeNames=[
        "ApproximateNumberOfMessages", "ApproximateNumberOfMessagesNotVisible"])["Attributes"]
    return int(attrs["ApproximateNumberOfMessages"]) + int(attrs["ApproximateNumberOfMessagesNotVisible"])


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--messages", type=int, default=2000)
    parser.add_argument("--workers", type=int, default=32)
    args = parser.parse_args()

    url = _sqs.get_queue_url(QueueName=QUEUE)["QueueUrl"]
    start = time.monotonic()
    with ThreadPoolExecutor(max_workers=args.workers) as pool:
        list(pool.map(lambda i: _sqs.send_message(QueueUrl=url, MessageBody=_body(i)), range(args.messages)))
    print(f"sent {args.messages} messages in {time.monotonic() - start:.2f}s", flush=True)

    initial = _depth(url)
    assert initial > 0, "queue empty right after burst; sends did not reach SQS"
    deadline = time.monotonic() + 120
    while time.monotonic() < deadline:
        remaining = _depth(url)
        print(f"queue depth: {remaining}", flush=True)
        if remaining == 0:
            print("PASS: queue fully drained by the consumer", flush=True)
            return
        time.sleep(3)
    print("WARNING: queue not fully drained within the window (LocalStack throughput ceiling)", flush=True)


if __name__ == "__main__":
    main()
